### PR TITLE
Refactor manual CPU/GPU dispatch in the tridiagonal solver

### PR DESCRIPTION
**Pull request by teonnik**

### CODE REVIEW

**Comment by teonnik** [comment]
bors try

**Comment by teonnik** [comment]
bors try

**Comment by teonnik** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by teonnik** [comment]
bors try

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 1]
```diff
@@ -139,26 +139,13 @@ inline SizeType problemSize(SizeType i_begin, SizeType i_end, const matrix::Dist
...
-      }
-      else {
-        initIndexTile(tile_row, index_tile.size().rows(), index_tile.ptr(), ts...);
-      }
-    };
-
-    di::transformDetach(di::Policy<DefaultBackend<D>::value>(), std::move(init_fn),
-                        index.readwrite_sender(tile_idx));
+    auto sender = di::whenAllLift(tile_row, index.readwrite_sender(tile_idx));
+    di::transformDetach(di::Policy<DefaultBackend<D>::value>(), initIndexTile_o, std::move(sender));
```
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:148]:
```suggestion
    di::transformDetach(di::Policy<DefaultBackend_v<D>>(), initIndexTile_o, std::move(sender));
```
and the same elsewhere.

Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:148]: This is not a request to change, but I just want to add the option which we use elsewhere:
```suggestion
    ex::start_detached(di::whenAllLift(tile_row, index.readwrite_sender(tile_idx)) |
      di::transform(di::Policy<DefaultBackend<D>::value>(), initIndexTile_o));
```
The `operator|` leads to some more template instantiations but is more consistent with other files. I'm indifferent about the choice so feel free to close this without action.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -203,14 +203,22 @@ add_library(
...
+target_add_warnings(dlaf.eigensolver)
+DLAF_addPrecompiledHeaders(dlaf.eigensolver)
+
+# Define DLAF's tridiagonal eigensolver library
+add_library(
+  dlaf.tridiagonal_eigensolver OBJECT
   eigensolver/tridiag_solver/mc.cpp
   eigensolver/tridiag_solver/kernels.cpp
   $<$<BOOL:${DLAF_WITH_GPU}>:eigensolver/tridiag_solver/gpu.cpp>
   $<$<BOOL:${DLAF_WITH_GPU}>:eigensolver/tridiag_solver/kernels.cu>
```
Comment (msimberg) [inline src/CMakeLists.txt:217]: For later consideration: split the kernels into more than one source file?

**Review by msimberg — CHANGES_REQUESTED** [review]

**Review by albestro — APPROVED** [review]
LGTM

I support what @msimberg already pointed out about `_v` as shortcut for `::values` together with the preference for the `operator|`.

As also said by him, nothing blocking or mandatory.

**Comment by teonnik** [comment]
bors try

File: include/dlaf/eigensolver/tridiag_solver/merge.h (hunk 1 above)
Comment (msimberg) [reply include/dlaf/eigensolver/tridiag_solver/merge.h]: We decided that this is ok as it is in the meeting.

**Review by msimberg — APPROVED** [review]

File: include/dlaf/eigensolver/tridiag_solver/impl.h [hunk 3]
```diff
@@ -76,43 +72,27 @@ std::vector<pika::shared_future<T>> cuppensDecomposition(Matrix<T, D>& mat_trd)
...
   for (SizeType i_split = 0; i_split < i_end; ++i_split) {
-    auto sender =
-        ex::when_all(mat_trd(LocalTileIndex(i_split, 0)), mat_trd(LocalTileIndex(i_split + 1, 0)));
-
-    // Cuppen's tridiagonal decomposition
-    offdiag_vals.push_back(
-        di::transform(di::Policy<DefaultBackend<D>::value>(), cuppensDecomp_o, std::move(sender)) |
-        ex::make_future());
+    offdiag_vals.push_back(cuppensDecompAsync<T, D>(mat_trd(LocalTileIndex(i_split, 0)),
+                                                    mat_trd(LocalTileIndex(i_split + 1, 0))));
```
Comment (albestro) [inline include/dlaf/eigensolver/tridiag_solver/impl.h:76]: I think you can use `readwrite_sender` for both occurrences of the access to `mat_trd`.

**Review by albestro — APPROVED** [review]

File: include/dlaf/eigensolver/tridiag_solver/impl.h (hunk 3 above)
Comment (teonnik) [reply include/dlaf/eigensolver/tridiag_solver/impl.h]: https://github.com/eth-cscs/DLA-Future/pull/647/commits/f4af88fb772005357257359da21ee7ee4aeeca72

**Pull request merged**